### PR TITLE
Custom theme template

### PR DESCRIPTION
**Pull request by crhallberg**
A guiding theme to get developers started on their own themes.

- [x] Reword tutorial to insist on copying this template.
- [x] `php index.php make_new_theme`
- [x] Finalize name of custom theme example that gets copied by generator.
- ~~Command line tools in `GETTING_STARTED.md`~~
- ~~Update wiki from `GETTING_STARTED.md`~~
- [x] Move `GETTING_STARTED.md` content to wiki.
- [x] Link to documentation from search/home.phtml.

### CODE REVIEW

File: config/vufind/config.ini [hunk 1]
```diff
@@ -58,15 +58,16 @@ theme           = bootprint3
 ; actual theme name.  For example, http://library.myuniversity.edu/vufind?ui=theme1
 ; would load the myTheme1 theme with the setting shown below.  Note that the values
 ; of "standard" and "mobile" are reserved for the default and mobile themes defined
-; above.
-;alternate_themes = theme1:myTheme1,theme2:myTheme2
+; above. Example: theme1:myTheme1,theme2:myTheme2
+alternate_themes = custom:custom_tmpl
```
Comment (demiankatz) [inline config/vufind/config.ini]: I assume that we don't really want to have this turned on by default....

File: themes/custom_tmpl/GETTING_STARTED.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Configuration
```
Comment (demiankatz) [inline themes/custom_tmpl/GETTING_STARTED.md]: This documentation is very helpful. However, before we merge, I think we should talk about our documentation strategy moving forward. Some of the information here overlaps with information in the wiki. We should probably try to avoid redundancy, or else we have more things that can potentially fall out of date and get ignored. Should we link from the wiki to here, or from here to the wiki? Should we think about best practices for what goes in the wiki (perhaps general information that changes infrequently) and what goes in GitHub (perhaps more detailed information that may be volatile -- this could give us a form of version-specific documentation).

Bottom line: we seem to be setting some sort of precedent here, so let's make sure we have a strategy before we go ahead with it. I like what you've done, though!

File: themes/custom_tmpl/templates/search/home.phtml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            }
+        });
+      });
+JS;
+  ?>
+  <?=$this->inlineScript(\Zend\View\Helper\HeadScript::SCRIPT, $ilsStatusScript, 'SET'); ?>
+  <div class="search-hero" role="search">
+    <?=$this->context($this)->renderInContext("search/searchbox.phtml", ['ignoreHiddenFilterMemory' => true])?>
+    <div class="image-credit">
+      Yosemite Valley. Photo by <a class="name" href="https://unsplash.com/@heytowner">John Towner</a>.
```
Comment (demiankatz) [inline themes/custom_tmpl/templates/search/home.phtml]: Is this definitely issued under an appropriate license? If so, do we need to cite a CC designation or something?

File: themes/custom_tmpl/templates/search/home.phtml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+<?
```
Comment (demiankatz) [inline themes/custom_tmpl/templates/search/home.phtml]: Copying the whole home.phtml file may cause long-term maintenance headaches -- I can easily see the core file changing and us forgetting to adjust this example. Is it possible to customize the template by rendering the parent version and wrapping things around it, or otherwise demonstrating customization with less duplication?

File: themes/custom_tmpl/GETTING_STARTED.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+There are three places in the VuFind configuration that you can change to see your custom theme in action.
+
+The first and most straight-forward is by changing `config.ini > [Site] > theme`. This will change the default theme for everyone who comes to your site and is the setting you should change for production.
+
+The two more development-oriented options are below `theme`: `alternate_themes` and `selectable_themes`. They both take comma-separated lists of themes in the format of `name1:theme1,name2:theme2`. Changing the theme via either of these two options sets a cookie that will keep your theme choice over the course of your session.
+
+Setting up themes under `alternate_themes` allows you to add a `ui=name` parameter to the end of your URL to see that theme. For example, the default configuration of `custom:custom_tmpl` would cause `https://vufind.org/?ui=custom` to show the theme described in the directory `custom_tmpl`.
```
Comment (demiankatz) [inline themes/custom_tmpl/GETTING_STARTED.md]: I'm assuming that we really don't want users editing this example theme -- it should be a template that they can copy. Perhaps we should actually treat this the same way that we treat custom code modules, and adjust install.php so that a user can request the creation of a theme, and the installer will actually copy the template theme to a new directory (user's name of choice). The unfortunate thing about that idea is that config.ini doesn't exist yet when install.php runs (it's created by the web-based installer later) so we can't easily turn on the user's custom theme by default without doing something a bit complicated.

File: config/vufind/config.ini (hunk 1 above)
Comment (crhallberg) [inline config/vufind/config.ini]: I thought this was a harmless way to see immediate results, but we can make it the commented-out example just the same.

File: themes/custom_tmpl/GETTING_STARTED.md (hunk 2 above)
Comment (crhallberg) [inline themes/custom_tmpl/GETTING_STARTED.md]: I see more and more communities putting their documentation in their code. I'd be happy to link from here to the wiki.

File: themes/custom_tmpl/templates/search/home.phtml (hunk 3 above)
Comment (crhallberg) [inline themes/custom_tmpl/templates/search/home.phtml]: I got the photo from an awesome new site [Unsplash](https://unsplash.com). [Everything on the site is under Copyright Zero](https://unsplash.com/license).

File: themes/custom_tmpl/GETTING_STARTED.md (hunk 5 above)
Comment (crhallberg) [inline themes/custom_tmpl/GETTING_STARTED.md]: You're right, I'll reword it.

File: themes/custom_tmpl/templates/search/home.phtml (hunk 4 above)
Comment (crhallberg) [inline themes/custom_tmpl/templates/search/home.phtml]: I chose search/home because it doesn't change very often and because it's effectively the homepage of VuFind. If any changes happen here, they're likely to be in the rows on the bottom (customized out) or in the searchbox itself (separate template).

File: config/vufind/config.ini (hunk 1 above)
Comment (demiankatz) [inline config/vufind/config.ini]: My main concern here is that I don't want to surprise people with hidden features when they go to production. Some people might be annoyed if they suddenly learned that their production site could be forced into a weird-looking demo theme through the addition of an arbitrary GET parameter. It's no big deal, obviously, but I think it's a little more professional to make it opt-in. 

Another consideration: directly referencing the template theme in config.ini may encourage people to edit it directly, which as discussed elsewhere is really not what we want people to do.

File: themes/custom_tmpl/GETTING_STARTED.md (hunk 2 above)
Comment (demiankatz) [inline themes/custom_tmpl/GETTING_STARTED.md]: I agree, I think that including more documentation with the code probably makes sense, especially given the discussions we've had about versioning documentation. I just want to be sure that if documentation in the code takes the place of documentation in the wiki, that we set up links accordingly so we don't have two versions of the same thing existing in parallel. This might be a good conversation for the wiki part of the next dev call.

File: themes/custom_tmpl/templates/search/home.phtml (hunk 3 above)
Comment (demiankatz) [inline themes/custom_tmpl/templates/search/home.phtml]: Excellent.

File: themes/custom_tmpl/GETTING_STARTED.md (hunk 5 above)
Comment (demiankatz) [inline themes/custom_tmpl/GETTING_STARTED.md]: Thanks!

File: themes/custom_tmpl/templates/search/home.phtml (hunk 4 above)
Comment (demiankatz) [inline themes/custom_tmpl/templates/search/home.phtml]: True. I think it's mainly the embedded AJAX call that gives me pause as something that could get messed up all too easily... but I don't really see a better way to handle it. Hmm.

File: themes/custom_tmpl/GETTING_STARTED.md (hunk 2 above)
Comment (crhallberg) [inline themes/custom_tmpl/GETTING_STARTED.md]: A good start would be to include links to appropriate wiki pages. That's what I intend to do when we close this request: move the content to the wiki and provide a link.

File: themes/custom_tmpl/GETTING_STARTED.md (hunk 5 above)
Comment (crhallberg) [inline themes/custom_tmpl/GETTING_STARTED.md]: My plan is to make a script that takes a name as a parameter. It then 
1. copies the custom_tmpl directory with that name
1. enables it as the theme
1. sets `selectable_themes` to `Bootstrap:bootstrap3,Modern:modern,$custom:$custom`.

Comment (demiankatz) [inline themes/custom_tmpl/GETTING_STARTED.md]: Makes sense to me. The only question is how best to do this:

a.) Stand-alone script like install.php
b.) Zend Framework controller action like the language tools, etc.
c.) Phing task

I think of those options I would choose b, since it makes it easiest for you to make use of existing VuFind tools... and Option a is probably the worst (I can't really see any good reason to suggest it, unless we actually build this into the existing install.php somehow, which is problematic for reasons I've previously described). Option c might make the file operations easier to implement, but every other aspect would be harder.

Perhaps you have already reached the same conclusion, or perhaps you have something else in mind. Either way, let me know if I can be of any assistance.

Comment (crhallberg) [inline themes/custom_tmpl/GETTING_STARTED.md]: Would it make sense to only include a zip file of the theme template? That way you must make a copy of the template to use it?

Comment (demiankatz) [inline themes/custom_tmpl/GETTING_STARTED.md]: I wouldn't go that far -- it would make managing the file in Git a pain. I think it's okay to simply make it clear that this is a template from the naming (and we could even try something more explicit like blank_custom_theme_template). If you wanted to take it a step further you could put the whole theme folder somewhere else in the directory tree.... but I think that might create as much confusion as it would prevent.

Comment (crhallberg) [inline themes/custom_tmpl/GETTING_STARTED.md]: I could change the name from `custom_tmpl` to `custom_theme_template` for sure.

Comment (demiankatz) [inline themes/custom_tmpl/GETTING_STARTED.md]: What about `custom_theme_example` -- that way, it's clear that it is an example, and not something that's meant to be used... and by removing the word "template" from the name, we might avoid confusion between the fact that this is a template for a theme, and the fact that themes contain templates (ahh, overlapping terminology). What do you think?

**Comment by crhallberg** [comment]
Is this just waiting for documentation?

**Comment by demiankatz** [comment]
It's on my list to test; should get to it shortly after the release.

**Comment by demiankatz** [comment]
@crhallberg, I've made a few minor adjustments:

- Renamed the "customtheme" generator to simply "theme" -- typing "generate theme" seems sufficiently expressive and saves a few keystrokes.
- Added some slightly more robust error checking to the file copy logic (check more return values, etc.).
- Refactored the custom search/home.phtml to include the bootstrap3 search/home.phtml, to reduce redundant logic and make future maintenance easier.

I think that leaves a few things that need to be done:

- Settle on a final name for the custom theme template. I'm wondering if we should rename `custom_theme_example` to `local_example_do_not_edit` or something like that. I think the use of "local" might help emphasize that this is how you build your local theme, and putting a "do_not_edit" in there will make people pause and look at the readme.
- We need to figure out our documentation strategy; do we want to keep the markdown file in the theme itself, or do we want to move more to the wiki? How do we want to integrate new documentation with existing things like https://vufind.org/wiki/development:architecture:less and https://vufind.org/wiki/development:architecture:user_interface?
- Once we have our documentation in order, we may want to edit the existing custom search/home.phtml to include a paragraph pointing the reader to the docs, to help them find their way.
